src/Raven.Server/Documents/ShardedTcpHandlers/ShardedSubscriptionConnection.cs

```diff
@@ -0,0 +1,328 @@
+            }
+            catch (Exception)
+            {
+                // ignored
+            }
+        }
+
+        protected override Task<string> OnClientAck(string subscriptionChangeVectorBeforeCurrentBatch)
+        {
+            return Task.FromResult<string>(null);
+        }
+
+        public static void SendShardedSubscriptionDocuments(ServerStore server, TcpConnectionOptions tcpConnectionOptions, JsonOperationContext.MemoryBuffer buffer)
+        {
+            var tcpConnectionDisposable = tcpConnectionOptions.ConnectionProcessingInProgress($"ShardedSubscription_{tcpConnectionOptions.ShardedContext.DatabaseName}");
+            try
+            {
+                var connection = new ShardedSubscriptionConnection(server, tcpConnectionOptions, tcpConnectionDisposable, buffer);
+                _ = connection.Run();
+            }
+            catch (Exception)
+            {
+                tcpConnectionDisposable?.Dispose();
+                throw;
+            }
+        }
+
+        private async Task StartShardSubscriptionWorkers()
+        {
+            AddToStatusDescription(CreateStatusMessage(ConnectionStatus.Create));
+
+            for (int i = 0; i < TcpConnection.ShardedContext.Count; i++)
+            {
+                var re = TcpConnection.ShardedContext.RequestExecutors[i];
+                if (re.Topology == null)
+                    await re.WaitForTopologyUpdate(re._firstTopologyUpdate);
+
+                var shard = TcpConnection.ShardedContext.GetShardedDatabaseName(i);
+
+                var shardWorker = new ShardedSubscriptionWorker(_options, shard, re);
+
+                var worker = new SubscriptionShardHolder
+                {
+                    Worker = shardWorker,
+                    PullingTask = shardWorker.RunInternal(CancellationTokenSource.Token),
+                    RequestExecutor = re
+                };
+
+                _shardWorkers.Add(shard, worker);
+            }
+
+            await WriteJsonAsync(new DynamicJsonValue
+            {
+                [nameof(SubscriptionConnectionServerMessage.Type)] = nameof(SubscriptionConnectionServerMessage.MessageType.ConnectionStatus),
+                [nameof(SubscriptionConnectionServerMessage.Status)] = nameof(SubscriptionConnectionServerMessage.ConnectionStatus.Accepted)
+            });
+        }
+
+        private async Task MaintainConnectionWithClientWorker()
+        {
+            var exceptions = new Dictionary<string, Exception>();
+            var shardsToReconnect = new List<string>();
+            var heartbeatSp = Stopwatch.StartNew();
+            var reconnectSp = Stopwatch.StartNew();
+            var didBatch = false;
+            while (CancellationTokenSource.IsCancellationRequested == false)
+            {
+                exceptions.Clear();
+                shardsToReconnect.Clear();
+                if (didBatch == false)
+                    await Task.Delay(500);
```

Comment:
Can we avoid this? Better to have a wait for event and a timeout.

Reply:
done